pandas/core/arrays/datetimelike.py

```diff
@@ -0,0 +1,152 @@
+# -*- coding: utf-8 -*-
+
+import numpy as np
+
+from pandas._libs import iNaT
+from pandas._libs.tslibs.timedeltas import delta_to_nanoseconds
+
+from pandas.tseries import frequencies
+
+import pandas.core.common as com
+from pandas.core.algorithms import checked_add_with_arr
+
+
+class DatetimeLikeArrayMixin(object):
+    """
+    Shared Base/Mixin class for DatetimeArray, TimedeltaArray, PeriodArray
+
+    Assumes that __new__/__init__ defines:
+        _data
+        freq
+    """
+
+    @property
+    def _box_func(self):
+        """
+        box function to get object from internal representation
+        """
+        raise com.AbstractMethodError(self)
+
+    def __iter__(self):
+        return (self._box_func(v) for v in self.asi8)
+
+    @property
+    def values(self):
+        """ return the underlying data as an ndarray """
+        return self._data.view(np.ndarray)
+
+    @property
+    def asi8(self):
+        # do not cache or you'll create a memory leak
+        return self.values.view('i8')
+
+    # ------------------------------------------------------------------
+    # Null Handling
+
+    @property  # NB: override with cache_readonly in immutable subclasses
+    def _isnan(self):
```

Comment:
this should be `isna()` ?

Reply:
I'm open to the name-change.  This is the existing name.